js/src/vm/PortableBaselineInterpret.cpp

```diff
@@ -3325,2718 +6031,2980 @@ PBIResult PortableBaselineInterpret(JSContext* cx_, State& state, Stack& stack,
+          PREDICT_NEXT(JumpTarget);
+          PREDICT_NEXT(LoopHead);
+        } else {
+          ADVANCE(JSOpLength_And);
+        }
+        DISPATCH();
+      }
+      CASE(Or) {
+        bool result;
+        Value v = VIRTSP(0).asValue();
+        if (v.isBoolean()) {
+          result = v.toBoolean();
+          NEXT_IC();
+        } else if (HybridICs) {
+          result = ToBoolean(SPHANDLE(0));
+          NEXT_IC();
+        } else {
+          IC_SET_ARG_FROM_STACK(0, 0);
+          IC_ZERO_ARG(1);
+          IC_ZERO_ARG(2);
+          INVOKE_IC(ToBool, false);
+          result = Value::fromRawBits(ic_ret).toBoolean();
+        }
+        int32_t jumpOffset = GET_JUMP_OFFSET(pc);
+        if (result) {
+          ADVANCE(jumpOffset);
+          PREDICT_NEXT(JumpTarget);
+          PREDICT_NEXT(LoopHead);
+        } else {
+          ADVANCE(JSOpLength_Or);
+        }
+        DISPATCH();
+      }
+      CASE(JumpIfTrue) {
+        bool result;
+        Value v = VIRTSP(0).asValue();
+        if (v.isBoolean()) {
+          result = v.toBoolean();
+          VIRTPOP();
+          NEXT_IC();
+        } else if (HybridICs) {
+          result = ToBoolean(SPHANDLE(0));
+          VIRTPOP();
+          NEXT_IC();
+        } else {
+          IC_POP_ARG(0);
+          IC_ZERO_ARG(1);
+          IC_ZERO_ARG(2);
+          INVOKE_IC(ToBool, false);
+          result = Value::fromRawBits(ic_ret).toBoolean();
+        }
+        int32_t jumpOffset = GET_JUMP_OFFSET(pc);
+        if (result) {
+          ADVANCE(jumpOffset);
+          PREDICT_NEXT(JumpTarget);
+          PREDICT_NEXT(LoopHead);
+        } else {
+          ADVANCE(JSOpLength_JumpIfTrue);
+        }
+        DISPATCH();
+      }
+      CASE(JumpIfFalse) {
+        bool result;
+        Value v = VIRTSP(0).asValue();
+        if (v.isBoolean()) {
+          result = v.toBoolean();
+          VIRTPOP();
+          NEXT_IC();
+        } else if (HybridICs) {
+          result = ToBoolean(SPHANDLE(0));
+          VIRTPOP();
+          NEXT_IC();
+        } else {
+          IC_POP_ARG(0);
+          IC_ZERO_ARG(1);
+          IC_ZERO_ARG(2);
+          INVOKE_IC(ToBool, false);
+          result = Value::fromRawBits(ic_ret).toBoolean();
+        }
+        int32_t jumpOffset = GET_JUMP_OFFSET(pc);
+        if (!result) {
+          ADVANCE(jumpOffset);
+          PREDICT_NEXT(JumpTarget);
+          PREDICT_NEXT(LoopHead);
+        } else {
+          ADVANCE(JSOpLength_JumpIfFalse);
+        }
+        DISPATCH();
+      }
+
+      CASE(Add) {
+        if (HybridICs) {
+          Value v0 = VIRTSP(0).asValue();
+          Value v1 = VIRTSP(1).asValue();
+          if (v0.isInt32() && v1.isInt32()) {
+            int64_t lhs = v1.toInt32();
+            int64_t rhs = v0.toInt32();
+            if (lhs + rhs >= int64_t(INT32_MIN) &&
+                lhs + rhs <= int64_t(INT32_MAX)) {
+              VIRTPOP();
+              VIRTSPWRITE(0, StackVal(Int32Value(int32_t(lhs + rhs))));
```

Comment:
Would there be benefit to storing the result of lhs+rhs in a local variable?
```suggestion
            int64_t res = v1.toInt32() + v0.toInt32();
            if (res >= int64_t(INT32_MIN) &&
                res <= int64_t(INT32_MAX)) {
              VIRTPOP();
              VIRTSPWRITE(0, StackVal(Int32Value(int32_t(res))));
```

Reply:
Sure, LLVM should do that in opt builds but it's reasonable for clarity anyway; done!